src/lib/instanceManager/injectWrapOnBeforeEventSend.js

```diff
@@ -0,0 +1,37 @@
+/*
+Copyright 2020 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+const clone = require("../utils/clone");
+
+const copyTo = (target, source) => {
+  Object.keys(target).forEach(key => {
+    delete target[key];
+  });
+  Object.keys(source).forEach(key => {
+    target[key] = source[key];
+  });
+};
+
+module.exports = ({ version, turbine }) => onBeforeEventSend => content => {
+  content.xdm.implementationDetails.name = `${content.xdm.implementationDetails.name}/reactor`;
+  content.xdm.implementationDetails.version = `${content.xdm.implementationDetails.version}+${version}`;
+  if (onBeforeEventSend) {
+    const contentClone = clone(content);
+    try {
+      onBeforeEventSend(contentClone);
+      copyTo(content.xdm, contentClone.xdm);
+      copyTo(content.data, contentClone.data);
+    } catch (e) {
+      turbine.logger.error(e);
+    }
```

Comment:
I think you're doing this cloning and try-catching because of [how Alloy handles errors coming from the `onBeforeEventSend` callback (it swallows the error and send the unmodified content)](https://github.com/adobe/alloy/blob/8f9b0975f9a5794964e242d6c51c9804e9d0a40b/src/core/createEvent.js#L63-L85)? I'm of the opinion that we should modify the way we handle errors coming from `onBeforeEventSend`, as I've described [here](https://github.com/adobe/alloy/pull/646#discussion_r543520623). I'd like to hear what you think about this.

If anything, a comment here describing why we need this cloning and try-catching would be useful. Also, I notice we're logging an error here if the customer's callback throws an error, but if Alloy were to log an error when the callback throws an error (which it probably should if we stick with the way Alloy is handling callback errors), then I don't think it would be necessary to log an error here, right?

Reply:
Yes, you are right, and I should add a comment here. I can't remember the original reason why we did the error catching like we do. I agree that we should probably just not send the hit in that case. If a customer wants to they can wrap their own code in a try catch and implement this logic. We should discuss this together though.

The reason why I'm not just letting the error go through is because that would cause Alloy to use the unmodified version of xdm thus removing the extension version.